File: 2021/deno/day-2.ts
const input = await Deno.readTextFile("./input/day-2.in");

const rows = input.split("\n");

let instructions = rows.map((row) => {
  const [key, value] = row.split(" ");

  return [key, Number(value)] as const;
});

/**
 * Part One
 */

const partOne = instructions.reduce(
  (prev, [dir, value]) => {
    switch (dir) {
      case "forward":
        return { ...prev, horizontal: prev.horizontal + value };
      case "down":
        return { ...prev, depth: prev.depth + value };
      case "up":
        return { ...prev, depth: prev.depth - value };
      default:
        return prev;
    }
  },
  { horizontal: 0, depth: 0 }
);

console.log("Part One:", partOne.horizontal * partOne.depth);

/**
 * Part Two
 */

const partTwo = instructions.reduce(
  (prev, [dir, value]) => {
    switch (dir) {
      case "forward":
        return {
          ...prev,
          horizontal: prev.horizontal + value,
          depth: prev.depth + prev.aim * value
        };
      case "down":
        return { ...prev, aim: prev.aim + value };
      case "up":
        return { ...prev, aim: prev.aim - value };
      default:
        return prev;
    }
  },
  { horizontal: 0, depth: 0, aim: 0 }
);

console.log("Part Two:", partTwo.horizontal * partTwo.depth);

File: 2021/swift/day-2.swift
import Foundation

let filename = "../input/day-2.in"
func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let instructions = contents.split(separator: "\n").map {
            $0.split(separator: " ")
        }.map {
            (String($0[0]), Int($0[1])!)
        }

        let partOne: ([(String, Int)]) -> Int = {
            var horizontal = 0
            var depth = 0

            for (instruction, units) in $0 {
                switch instruction {
                case "forward":
                    horizontal += units
                case "down":
                    depth += units
                case "up":
                    depth -= units
                default:
                    print("Unknown instruction")
                }
            }

            return horizontal * depth
        }

        print("Part one: \(partOne(instructions))")

        let partTwo: ([(String, Int)]) -> Int = {
            var horizontal = 0
            var depth = 0
            var aim = 0

            for (instruction, units) in $0 {
                switch instruction {
                case "forward":
                    horizontal += units
                    depth += (aim * units)
                case "down":
                    aim += units
                case "up":
                    aim -= units
                default:
                    print("Unknown instruction")
                }
            }

            return horizontal * depth
        }

        print("Part two: \(partTwo(instructions))")
    } catch {
        print("Failed to read input")
    }
}

main()
